ENH: Allow safe access to `.book` in ExcelWriter

rhshadrach opened the pull request:
Towards #45572 

- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

The plan to expose both `sheets` and `book` as public attributes of `ExcelWriter`. We need to make sure that `sheets` is correct if the user modifies the state via accessing `book`.

This accesses a protected attribute for xlwt; I looked at the xlwt code and there is no public access to determine the sheets. The xlwt repository is read-only, we only support the most recent version, and xlwt will be removed in pandas 2.0. So I think this is okay.

rhshadrach committed:
ENH: Allow safe access to `.book` in ExcelWriter

rhshadrach committed:
Add GH # in tests

phofl commented:
This Looks related?

rhshadrach committed:
Use context manager in tests

rhshadrach commented:
Thanks @phofl - wasn't using a context manager for ExcelWriter

rhshadrach commented:
Failure is unrelated.

```
FAILED pandas/tests/computation/test_eval.py::TestEval::test_unary_in_array[python-python]
FAILED pandas/tests/computation/test_eval.py::TestEval::test_unary_in_array[python-pandas]
FAILED pandas/tests/computation/test_eval.py::TestOperations::test_column_in[python-python]
FAILED pandas/tests/computation/test_eval.py::TestOperations::test_column_in[python-pandas]
```

phofl commented on pandas/io/excel/_openpyxl.py:
```diff
@@ -440,7 +442,6 @@ def write_cells(
                     target_index = self.book.index(old_wks)
                     del self.book[sheet_name]
                     wks = self.book.create_sheet(sheet_name, target_index)
-                    self.sheets[sheet_name] = wks
```
Why can we remove this? Does create_sheet modify self.book?

rhshadrach replied:
That's correct

```
from openpyxl import load_workbook

book = load_workbook("test.xlsx")
print({name: book[name] for name in book.sheetnames})
book.create_sheet("test_sheet", 0)
print({name: book[name] for name in book.sheetnames})
```

gives

```
{'Sheet1': <Worksheet "Sheet1">}
{'test_sheet': <Worksheet "test_sheet">, 'Sheet1': <Worksheet "Sheet1">}
```

phofl replied:
Thx

phofl commented on pandas/io/excel/_xlwt.py:
```diff
@@ -63,6 +63,11 @@ def __init__(
 
+    @property
+    def sheets(self) -> dict[str, Any]:
+        result = {sheet.name: sheet for sheet in self.book._Workbook__worksheets}
```
Could directly return the dict like above

rhshadrach replied:
In general I like having the return line separate as it gives a variable to inspect allowing for easier debugging. Actually meant to do this above. I'll change the above for now, but if others don't like this style I have no objection to having the return all on one line.

phofl replied:
APart from consistency no strong opinion on my side

phofl reviewed (commented):
Small comments

phofl commented:
You did not add a test for xlsxwriter, was this on purpose?

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into sheets_property

rhshadrach committed:
Add back xlsxwriter sheets, add tests

rhshadrach commented:
> You did not add a test for xlsxwriter, was this on purpose?

Yes, but I realize now that was a mistake. I removed the `sheets` attribute from the writer for xlsxwriter because it was never used (hence, no tests). But that creates an inconsistency between the writers, so I've added it back on.

rhshadrach committed:
Add sheets as an abstract property

jreback commented on pandas/io/excel/_odswriter.py:
```diff
@@ -58,6 +58,16 @@ def __init__(
         self._style_dict: dict[str, str] = {}
 
+    @property
+    def sheets(self) -> dict[str, Any]:
```
is there a way to type these more specifically?

rhshadrach replied:
Not easily, as far as I can tell. I plan to look into this more in the future.

Regarding whatsnew - this is not user facing. Currently in docs no attribute of ExcelWriter is public. I'd like to refactor ExcelWriter (namely, making protected attributes start with `_`) before updating docs.

jreback replied:
got it totally fine (make issues if you cannot to get to it soon), otherwise PR's totally fine.

> is there a way to type these more specifically?

can we use the base class generic for typing? (again nbd)

jreback reviewed (changes requested):
can you add a whatsnew
can you add to the api doc reference
these also need a doc-string (the properties) with a versionadded 1.5

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into sheets_property

rhshadrach committed:
docstrings

rhshadrach commented:
@jreback  - added docstrings. I'd like to hold off updating the docs until the API is ready to be public.

jreback commented on pandas/io/excel/_xlsxwriter.py:
```diff
@@ -222,11 +227,9 @@ def write_cells(
-        if sheet_name in self.sheets:
-            wks = self.sheets[sheet_name]
-        else:
+        wks = self.book.get_worksheet_by_name(sheet_name)
```
i would change `.book` to be a private attribute `._book` and then expose the `.book` property (followups for sure)

rhshadrach replied:
Would the property be something besides just `return self._book`? In other words, why a property?

jreback reviewed (approved):
comments for followup
